Python/bltinmodule.c

```diff
@@ -2667,15 +2685,73 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 int overflow;
                 value = PyLong_AsLongAndOverflow(item, &overflow);
                 if (!overflow) {
-                    f_result += (double)value;
+                    f_result.sum += (double)value;
                     Py_DECREF(item);
                     continue;
                 }
             }
-            if (c && isfinite(c)) {
-                f_result += c;
+            _csum_neumaier_finalize(&f_result);
+            result = PyFloat_FromDouble(f_result.sum);
+            if (result == NULL) {
+                Py_DECREF(item);
+                Py_DECREF(iter);
+                return NULL;
+            }
+            temp = PyNumber_Add(result, item);
+            Py_DECREF(result);
+            Py_DECREF(item);
+            result = temp;
+            if (result == NULL) {
+                Py_DECREF(iter);
+                return NULL;
+            }
+        }
+    }
+
+    if (PyComplex_CheckExact(result)) {
+        Py_complex z = PyComplex_AsCComplex(result);
+        _csum cr_result = {z.real, 0.0};
+        _csum ci_result = {z.imag, 0.0};
+        Py_SETREF(result, NULL);
+        while (result == NULL) {
+            item = PyIter_Next(iter);
+            if (item == NULL) {
+                Py_DECREF(iter);
+                if (PyErr_Occurred()) {
+                    return NULL;
+                }
+                _csum_neumaier_finalize(&cr_result);
+                _csum_neumaier_finalize(&ci_result);
+                return PyComplex_FromDoubles(cr_result.sum, ci_result.sum);
+            }
+            if (PyComplex_CheckExact(item)) {
+                z = PyComplex_AsCComplex(item);
+                _csum_neumaier_step(&cr_result, z.real);
+                _csum_neumaier_step(&ci_result, z.imag);
+                Py_DECREF(item);
+                continue;
+            }
+            if (PyLong_Check(item)) {
+                long value;
+                int overflow;
+                value = PyLong_AsLongAndOverflow(item, &overflow);
+                if (!overflow) {
+                    cr_result.sum += (double)value;
+                    ci_result.sum += 0.0;
+                    Py_DECREF(item);
+                    continue;
+                }
+            }
+            if (PyFloat_Check(item)) {
+                double value = PyFloat_AS_DOUBLE(item);
+                cr_result.sum += value;
+                ci_result.sum += 0.0;
+                Py_DECREF(item);
```

Comment:
You should put `_Py_DECREF_SPECIALIZED(item, _PyFloat_ExactDealloc);` here.

Reply:
Yep, forgot to add that in above commit.